server/src/state/system.rs

```diff
@@ -0,0 +1,529 @@
+use iggy::models::user_status::UserStatus;
+use iggy::partitions::create_partitions::CreatePartitions;
+use iggy::partitions::delete_partitions::DeletePartitions;
+use iggy::personal_access_tokens::delete_personal_access_token::DeletePersonalAccessToken;
+use iggy::streams::create_stream::CreateStream;
+use iggy::streams::delete_stream::DeleteStream;
+use iggy::streams::update_stream::UpdateStream;
+use iggy::topics::create_topic::CreateTopic;
+use iggy::topics::delete_topic::DeleteTopic;
+use iggy::topics::update_topic::UpdateTopic;
+use iggy::users::change_password::ChangePassword;
+use iggy::users::create_user::CreateUser;
+use iggy::users::delete_user::DeleteUser;
+use iggy::users::update_permissions::UpdatePermissions;
+use iggy::users::update_user::UpdateUser;
+use iggy::utils::expiry::IggyExpiry;
+use iggy::utils::timestamp::IggyTimestamp;
+use iggy::utils::topic_size::MaxTopicSize;
+use std::collections::HashMap;
+use std::fmt::Display;
+use tracing::{debug, error};
+
+#[derive(Debug)]
+pub struct SystemState {
+    pub streams: HashMap<u32, StreamState>,
+    pub users: HashMap<u32, UserState>,
+}
+
+#[derive(Debug)]
+pub struct StreamState {
+    pub id: u32,
+    pub name: String,
+    pub created_at: IggyTimestamp,
+    pub topics: HashMap<u32, TopicState>,
+    pub current_topic_id: u32,
+}
+
+#[derive(Debug)]
+pub struct TopicState {
+    pub id: u32,
+    pub name: String,
+    pub partitions: HashMap<u32, PartitionState>,
+    pub consumer_groups: HashMap<u32, ConsumerGroupState>,
+    pub compression_algorithm: CompressionAlgorithm,
+    pub message_expiry: IggyExpiry,
+    pub max_topic_size: MaxTopicSize,
+    pub replication_factor: Option<u8>,
+    pub created_at: IggyTimestamp,
+    pub current_consumer_group_id: u32,
+}
+
+#[derive(Debug)]
+pub struct PartitionState {
+    pub id: u32,
+    pub created_at: IggyTimestamp,
+}
+
+#[derive(Debug)]
+pub struct PersonalAccessTokenState {
+    pub name: String,
+    pub token_hash: String,
+    pub expiry_at: Option<IggyTimestamp>,
+}
+
+#[derive(Debug)]
+pub struct UserState {
+    pub id: u32,
+    pub username: String,
+    pub password_hash: String,
+    pub status: UserStatus,
+    pub permissions: Option<Permissions>,
+    pub personal_access_tokens: HashMap<String, PersonalAccessTokenState>,
+}
+
+#[derive(Debug)]
+pub struct ConsumerGroupState {
+    pub id: u32,
+    pub name: String,
+}
+
+// TODO: Consider handling stream and topic purge
+impl SystemState {
+    pub async fn init(entries: Vec<StateEntry>) -> Result<Self, IggyError> {
+        let mut streams = HashMap::new();
+        let mut users = HashMap::new();
+        let mut current_stream_id = 0;
+        let mut current_user_id = 0;
+        for entry in entries {
+            debug!(
+                "Processing state entry with code: {}, name: {}",
+                entry.code,
+                get_name_from_code(entry.code).unwrap_or("invalid_command")
+            );
+            match entry.code {
```

Comment:
I wonder if we could maybe make the `code` field be an enum and have it encapsulate all of that complexity inside of it's methods. To convert it back to an u32 for persistence purposes we could use `From` trait, maybe thats a `good first issue` for new Aaron :see_no_evil:

Reply:
Yes, I was also thinking that we could even have some sort of  `Command` trait, that would contain `get_code()` method.

Reply:
Fixed